src/net/tcp/listener.rs

```diff
@@ -48,7 +60,26 @@ impl TcpListener {
     /// established, the corresponding [`TcpStream`] and the remote peer’s
     /// address will be returned.
     pub async fn accept(&self) -> Result<(TcpStream, SocketAddr)> {
+        poll_fn(|cx| self.poll_accept(cx)).await
+    }
+
+    /// Polls to accept a new incoming connection to this listener.
+    ///
+    /// If there is no connection to accept, Poll::Pending is returned and the current task will be
+    /// notified by a waker. Note that on multiple calls to poll_accept, only the Waker from the
+    /// Context passed to the most recent call is scheduled to receive a wakeup.
+    pub fn poll_accept(&self, cx: &mut Context<'_>) -> Poll<Result<(TcpStream, SocketAddr)>> {
         loop {
+            {
+                let mut lock = self.accept_fut.lock().unwrap();
```

Comment:
Agreed. I think this ugly pattern will keep showing up, for basically all `poll_` implementation. Maybe we should go ahead and refactor the code to work nicely for poll_ functions, and then straightforwardly build the async version on top of them?

Reply:
Two more ideas:

/1/ Use `unimplemented!(...)` for `poll_` methods. It'd be good to hear more about the use case that needs polling variants.

/2/ Stash the poll state in the host itself. This would avoid the need for a mutex. We can't take the above approach directly as we can't hold onto the "world" across an await, but we could potentially grab the waker from the cx passed to poll, stash it with an identifier for the listener, and then _wake_ once a syn arrives.